Initial import of Anki flashcard app

dsd opened the pull request:
Anki is a popular app for remembering things through the use of flashcards. It would be great to have it available via flathub. Here is my attempt at packaging it.

cc original app author @dae

TingPing commented on net.ankiweb.Anki.json:
```diff
@@ -0,0 +1,103 @@
+  "command": "anki",
+  "finish-args": [
+     "--socket=x11",
+     "--share=network"
```
You'll want `--socket=wayland`, `--share=ipc`, and `--device=dri` also.

dsd commented:
Anki would normally install mime type associations via https://github.com/dae/anki/blob/master/anki.xml
```
	xdg-mime install anki.xml --novendor
	xdg-mime default anki.desktop application/x-anki
	xdg-mime default anki.desktop application/x-apkg
```
How can I integrate this correctly in the manifest? Currently I remove those lines from the Makefile.

Also the build process complains confusingly about the icon:
```
Renaming anki.desktop to net.ankiweb.Anki.desktop
Committing stage cleanup to cache
Finishing app
Exporting share/applications/net.ankiweb.Anki.desktop
Exporting share/icons/net.ankiweb.Anki.png
Please review the exported files and the metadata
Committing stage finish to cache
Exporting net.ankiweb.Anki to repo
WARNING: Icon referenced in desktop file but not exported: net.ankiweb.Anki
```

It complains that it was not exported 4 lines after having exported it?

TingPing commented on net.ankiweb.Anki.json:
```diff
@@ -0,0 +1,103 @@
+        "sed -i -e 's/Icon=anki/Icon=net.ankiweb.Anki/' anki.desktop",
+        "sed -i -e 's:/usr/share:/app/share:' tools/runanki.system",
+        "sed -i -e '/xdg-mime/d' Makefile",
+        "make install PREFIX=/app",
```
You can patch it here first but it would be great if it made it upstream eventually.

TingPing commented:
> Anki would normally install mime type associations via https://github.com/dae/anki/blob/master/anki.xml

Flatpak's can still export mime assocations the installed file just needs to be renamed to the app-id like everything else (make not of icon references in them also).

The `xdg-mime` tool not being there might be annoying, maybe just package it as a build dep, maybe just patch it to directly `install` the file (I don't think `xdg-mime install` does anything of value)

> It complains that it was not exported 4 lines after having exported it?

Because `Exporting share/icons/net.ankiweb.Anki.png` is the wrong path. `share/icons/hicolor/$ICON_SIZE/apps/net.ankiweb.Anki.png` is correct.

TingPing commented on net.ankiweb.Anki.json:
```diff
@@ -64,7 +64,8 @@
       "build-commands": [ 
         "python3 configure.py --assume-shared --confirm-license --no-designer-plugin --no-qml-plugin --sysroot=/app -w --sip=/app/bin/sip --sip-incdir=/app/include -verbose QMAKE_CFLAGS_RELEASE='-I/usr/include/python3.5m/' QMAKE_CXXFLAGS_RELEASE='-I/usr/include/python3.5m/'", 
-        "make -j3", "make install DESTDIR=/app"
+        "make",
```
Well, `make` just defaults to a single job. Do `make -j \`nproc\`` to use available threads.

TingPing commented:
bot, build net.ankiweb.Anki

dsd commented:
Is the ARM build failure anything to worry about? It's not clear what it is failing on.

I'll try to upstream Makefile fixes to make it more suitable here. https://github.com/dae/anki/pull/227

barthalion commented on bs4.json:
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "bs4",
```
Please prefix python libraries with `python-`. It's not a mistake per se but I have a script that looks for commonly used modules.

TingPing replied:
These are all generated files for the python deps

barthalion replied:
Too bad. I'll submit pull request for this later.

dsd replied:
Yes, `flatpak-pip-generator` made those files for me.

TingPing replied:
https://github.com/flatpak/flatpak-builder-tools/commit/47340df0c5a8625ae6a20ab1db96bdaa2f323105

barthalion commented:
bot, build net.ankiweb.Anki

barthalion commented:
Also @dae, we will be more than happy to give you push access if you would like to have it.

barthalion commented:
Build failure on ARM smells like race condition between installing file and creating directory where it should be installed. Try passing `-j1` to `make install`.

dsd commented:
The makefile is just a simple `install` recipe so I don't see that there is anything that could be racing.
https://github.com/dae/anki/blob/master/Makefile#L8

I can't see the github complaint about failing builds anymore. Does that mean that it succeeded upon rebuild?

barthalion commented:
No, it's issue with builders not reporting back the results. I will try to reproduce the failure locally.

dsd commented:
Ah, the failure is buried higher up in the log:
```
'designer/icons/media-record.png' -> '/app/share/anki/designer/icons/media-record.png'
cp: failed to preserve ownership for /app/share/anki/designer/icons/media-record.png: No such file or directory
```

but that's very weird indeed. Is the fs on the ARM builder corrupt or something?

barthalion commented on net.ankiweb.Anki.json:
```diff
@@ -0,0 +1,117 @@
+{
+  "app-id": "net.ankiweb.Anki",
+  "runtime": "org.kde.Platform",
+  "runtime-version": "5.9",
```
Btw, 5.10 is out.

dsd replied:
Anki requires 5.9.x
https://github.com/dae/anki/blob/master/aqt/qt.py#L43

barthalion commented:
> Is the fs on the ARM builder corrupt or something?
It's fishy that the error always happen on the same file though.

dsd commented:
>  It's fishy that the error always happen on the same file though.

Have you confirmed this? I can only see a single build on https://flathub.org/builds/#/builders/5, not sure if we have seen this error multiple times

dsd commented:
`media-record.png` is a symlink whereas the other files in that directory are regular files. If it is a consistent failure, then maybe that has something to do with it.

barthalion commented:
https://flathub.org/builds/#/builders/78

dsd commented:
So what do you suggest as next steps here? Maybe one of the following?

* In the build manifest, delete the symlink before copying and then recreate it after using `ln -s`?
* Give me access to the ARM builder so I can login and investigate why it can't copy a symlink even though all the other builders are fine?
* Ship initially with no ARM build?

I think I've addressed all the review comments but please let me know if I'm missing anything else.

nedrichards commented:
@dsd @ramcq can help you with the ARM builders.

ramcq commented:
I've seen this error before in a flatpak build but hard to dig around for it from my phone. It was maybe something to do with trying to cp -a a symlink rather than just cp -r? I can have a check tomorrow.

ramcq commented:
Yeah - https://github.com/flathub/org.freedesktop.Sdk.Extension.mono5/commit/e25f6b572ba5eab85c4eb1f1286b8301d3fd8253 or https://github.com/flathub/org.libreoffice.LibreOffice/commit/41b361bf4973b0981f834fe58b3f808b0ca81634 - maybe there's something weird about the kernel or FS on those ARM builders?

barthalion commented:
@dsd Any chance you could look at fixing that `cp -a` invocation or disable `arm` for now?

ramcq commented:
@barthalion There's no need to disable arm here, it's very easy to avoid this problem...

barthalion commented:
Which I ask for now or later, when submitter can use `bot, build` himself.

dsd commented:
pushed a change that should avoid the issue there

nedrichards commented:
bot, build net.ankiweb.Anki

nedrichards commented:
bot, build net.ankiweb.Anki

nedrichards commented:
Repository has been created: https://github.com/flathub/net.ankiweb.Anki
